src/LoanAgent.sol

```diff
@@ -49,73 +38,166 @@ contract LoanAgent {
         Terminated
     }
 
+    AddressInfo public addressInfo;
     NodeStatus public status;
+    LoanAgentInfo public loanAgentInfo;
+
+    /// checkpoint => withdrawnAmount
+    mapping(uint => uint) withdrawHistory;
+    /// checkpoint => available balance after withdrawal
+    mapping(uint => uint) availableBalanceHistory;
+    // checkpoint => miner share
+    mapping(uint => uint) ownerWithrawableAmount;
+    // total amount of miner share accummulated
+    uint totalOwnerCredit;
+    uint lastWithdrawlCheckpoint;
 
-    // Stakepool factory
     address public immutable factory;
+    address public immutable aFil;
 
-    // should be called using LoanAgentFactory
     constructor(
         AddressInfo memory _addressInfo,
         LoanAgentInfo memory _loanAgentInfo,
-        address _factory
+        address _factory,
+        address _aFil
     ) {
-        // Call ChangeWorkerAddress in MinerAPI: https://docs.zondax.ch/fevm/filecoin-solidity/api/actors/Miner#changeworkeraddress
+        addressInfo = _addressInfo;
+        loanAgentInfo = _loanAgentInfo;
         factory = _factory;
+        aFil = _aFil;
+        status = NodeStatus.InQueue;
     }
 
-    modifier onlyFactory {
+    modifier onlyFactory() {
+        require(msg.sender == factory, "can only be called by factory");
         _;
     }
 
-    modifier onlyOracle {
+    modifier onlyOracle() {
```

Comment:
addressInfo.oracle is address of oracle node associated with the LoanAgent i.e hosted by the LoanAgent. onlyOracle modifier checks call from oracle contract. The address can be stored directly or pulled from loan agent factory